src/EntityComponentManager_TEST.cc

```diff
@@ -2009,6 +2013,198 @@ TEST_P(EntityComponentManagerFixture, SetChanged)
       manager.ComponentState(e2, c2.first));
 }
 
+//////////////////////////////////////////////////
+TEST_P(EntityComponentManagerFixture, SerializedStateMapMsgAfterRemoveComponent)
+{
+  // Create entity
+  Entity e1 = manager.CreateEntity();
+  manager.CreateComponent<IntComponent>(e1, IntComponent(123));
+  auto e1c1 =
+    manager.CreateComponent<DoubleComponent>(e1, DoubleComponent(0.0));
+  auto e1c2 =
+    manager.CreateComponent<StringComponent>(e1, StringComponent("int"));
+
+  manager.RemoveComponent(e1, e1c1);
+  manager.RemoveComponent(e1, e1c2);
+
+  // Serialize into a message
+  msgs::SerializedStateMap stateMsg;
+  manager.State(stateMsg);
+
+  // Check message
+  {
+    auto iter = stateMsg.entities().find(e1);
+    const auto &e1Msg = iter->second;
+    auto compIter = e1Msg.components().begin();
+
+    // First component
+    const auto &e1c2Msg = compIter->second;
+    compIter++;
+    EXPECT_TRUE(e1c2Msg.remove());
```

Comment:
This expectation is failing on macOS, as well as line 2052:

https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-homebrew-amd64/3546/testReport/(root)/EntityComponentManagerRepeat_EntityComponentManagerFixture/SerializedStateMapMsgAfterRemoveComponent_1/

I'm not sure yet what could be different there.

Reply:
It must be the way I access the components with `compIter`. It seems that the order they are accessed is different in macOS than Linux. This has been solved [like this](https://github.com/ignitionrobotics/ign-gazebo/blob/master/src/EntityComponentManager_TEST.cc#L1528-L1554) in the `State` test. Should I do the same?

Reply:
The order of elements in `components()` seems like an implementation detail that we probably don't want to rely on because it depends on the (undetermined) order in `removedComponents` which is an `std::unordered_multimap`. If we want to preserve the order of components, you can use `std::unordered_map<Entity, std::vector<ComponentKey>>` for `removedComponents` instead.

A different solution would be to have the expectations in a map from `component type id` -> `bool expectation` when creating the components at the beginning of the test. The component type id is returned by `CreateComponent` as the first element of a pair. Once you have the map, you can iterate through the components in `stateMsg` and look them up from the map based on their type.